src/index.js

```diff
@@ -32,20 +49,34 @@ const state = {};
 
 function forEachListener(
   props: Props,
-  iteratee: (eventName: string, listener: Function, capture?: boolean) => any
+  iteratee: (eventName: string, listener: Function, options?: EventOptions) => any
 ): void {
   for (const name in props) {
-    if (name.substring(0, 2) === 'on' && props[name] instanceof Function) {
-      const capture = name.substr(-7).toLowerCase() === 'capture';
+    if (name.substring(0, 2) !== 'on') continue;
+
+    const prop = props[name];
+    const type = typeof prop;
+    const isObject = type === 'object';
+    const isFunction = type === 'function';
+
+    if (!isObject && !isFunction) continue;
 
-      let eventName = name.substring(2).toLowerCase();
-      eventName = capture ? eventName.substring(0, eventName.length - 7) : eventName;
+    const capture = name.substr(-7).toLowerCase() === 'capture';
+    let eventName = name.substring(2).toLowerCase();
+    eventName = capture ? eventName.substring(0, eventName.length - 7) : eventName;
 
-      iteratee(eventName, props[name], capture);
+    if (isObject) {
+      iteratee(eventName, prop.handler, prop.options);
+    } else {
+      iteratee(eventName, prop, assign({}, defaultEventOptions, {capture}));
     }
   }
 }
 
+export function withOptions(handler: Function, options: EventOptions): {handler: Function, options: EventOptions} {
+  return {handler, options: assign({}, defaultEventOptions, options)};
```

Comment:
The presence of certainly the key is I think that it is not necessary to check.
Following is of such a simple check only looks good.

```javascript
import warning from 'warning';

// ...

export function withOptions(/* ... */) {
  if (process.env.NODE_ENV !== 'production' && typeof options === 'undefined') {
    warning(options, '...');
  }
  // ...
}
```